pkg/config/crd/store.go

```diff
@@ -0,0 +1,374 @@
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	// import GKE cluster authentication plugin
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	// import OIDC cluster authentication plugin, e.g. for Tectonic
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
+
+	"istio.io/mixer/pkg/config/store"
+)
+
+const (
+	// API group / version for istio config.
+	apiGroup        = "config.istio.io"
+	apiVersion      = "v1alpha2"
+	apiGroupVersion = apiGroup + "/" + apiVersion
+)
+
+const (
+	// defaultResyncPeriod is the resync period for the k8s cache.
+	// TODO: allow customization.
+	defaultResyncPeriod = time.Minute
+
+	// initializationPeriod is the duration to wait for getting
+	// initial data. Mixer doesn't want to get added events for
+	// initial data since it may arrive in any order, and it
+	// would result in temporary invalid status.
+	initializationPeriod = time.Second * 2
+)
+
+// Store offers store.Store2 interface through kubernetes custom resource definitions.
+type Store struct {
+	conf      *rest.Config
+	ns        map[string]bool
+	caches    map[string]cache.Store
+	scheme    *runtime.Scheme
+	kinds     map[string]proto.Message
+	resources map[string]metav1.APIResource
+	client    *rest.RESTClient
+
+	mu  sync.Mutex
+	eqs map[string][]*eventQueue
+}
+
+var _ store.Store2 = &Store{}
+
+// SetValidator implements store.Store2 interface.
+func (s *Store) SetValidator(store.Validator) {
+	glog.Errorf("Validator is not implemented yet")
+}
+
+// RegisterKind implements store.Store2 interface.
+func (s *Store) RegisterKind(kind string, spec proto.Message) {
+	s.scheme.AddKnownTypeWithName(
+		schema.GroupVersionKind{Group: apiGroup, Version: apiVersion, Kind: kind},
+		&resource{},
+	)
+	s.scheme.AddKnownTypeWithName(
+		schema.GroupVersionKind{Group: apiGroup, Version: apiVersion, Kind: kind + "List"},
+		&resourceList{},
+	)
+	s.kinds[kind] = spec
+}
+
+// Init implements store.Store2 interface.
+func (s *Store) Init(ctx context.Context) error {
+	var err error
+	s.conf.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: serializer.NewCodecFactory(s.scheme)}
+	s.client, err = rest.RESTClientFor(s.conf)
+	if err != nil {
+		return err
+	}
+	d := discovery.NewDiscoveryClientForConfigOrDie(s.conf)
+	resources, err := d.ServerResourcesForGroupVersion(apiGroupVersion)
+	if err != nil {
+		return err
+	}
+	dyn, err := dynamic.NewClient(s.conf)
+	if err != nil {
+		return err
+	}
+	s.caches = make(map[string]cache.Store, len(s.kinds))
+	for _, res := range resources.APIResources {
+		if _, ok := s.kinds[res.Kind]; ok {
+			// This is a bit tricky; the variable "res" may refer to the same address
+			// whose content can be updated through the iteration, and that goes
+			// the resource client to lookup different resources. It's safer to make a
```

Comment:
That goes `to` the resource client  ?  'to' is missing.
Can you specify which line may modify the contents of res?